framework/src/kernels/ArrayDiffusion.C

```diff
@@ -0,0 +1,81 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "ArrayDiffusion.h"
+
+registerMooseObject("MooseApp", ArrayDiffusion);
+
+template <>
+InputParameters
+validParams<ArrayDiffusion>()
+{
+  InputParameters params = validParams<ArrayKernel>();
+  params.addParam<MaterialPropertyName>("diffusion_coefficient", "The name of the diffusivity");
+  MooseEnum opt("scalar=0 array=1 full=2", "array");
+  params.addParam<MooseEnum>("diffusion_coefficient_type", opt, "Diffusion coefficient type");
+  params.addClassDescription(
+      "The array Laplacian operator ($-\\nabla \\cdot \\nabla u$), with the weak "
+      "form of $(\\nabla \\phi_i, \\nabla u_h)$.");
+  return params;
+}
+
+ArrayDiffusion::ArrayDiffusion(const InputParameters & parameters)
+  : ArrayKernel(parameters), _dc_type(getParam<MooseEnum>("diffusion_coefficient_type"))
+{
+  if (_dc_type == 0)
+    _d = &getMaterialProperty<Real>("diffusion_coefficient");
+  else if (_dc_type == 1)
+    _d_array = &getMaterialProperty<RealArrayValue>("diffusion_coefficient");
+  else if (_dc_type == 2)
+    _d_2d_array = &getMaterialProperty<RealArray>("diffusion_coefficient");
```

Comment:
I looked at these typedefs in `MooseTypes` and it's not immediately clear to me the motivation behind their names. I look at `RealArray` and presume that it's an array of `Reals`. Is that right? With the enumeration description of `full`, I'm guessing that it's not. And what is `RealArrayValue` supposed to mean?

Reply:
RealArrayValue is actually an Eigen vector type, equivalent with `VectorXd`, where `X` stands for dynamic, `d` stands for double. `RealArray` is actually an Eigen matrix, `MatrixXd`. I have to admit that I came up this name without too much thinking, so I am willing to accept a better name. Or I can change it to `RealMatrixValue`, what do you think?

Reply:
Hmm...we're starting to run out of names that don't clash with existing names. I personally like `RealEigenVector` and `RealEigenMatrix`. @friedmud likes names; maybe he has a suggestion.

Reply:
I do not quite `Eigen` in the name because we also used it for eigenvalue. It can cause confusion although we know here `Eigen` is the package name. The underlining data structures for vector and matrix in Eigen are the same. I do not have good ideas now for naming these two ;-) I will need the final for changing the code.

Reply:
Since I do not have better ideas, I will make this change, I like these names of differentiating vector and matrix.